src/vs/base/node/powershell.ts

```diff
@@ -0,0 +1,324 @@
+const IntRegex: RegExp = /^\d+$/;
+
+const PwshMsixRegex: RegExp = /^Microsoft.PowerShell_.*/;
+const PwshPreviewMsixRegex: RegExp = /^Microsoft.PowerShellPreview_.*/;
+
+// The platform details descriptor for the platform we're on
+const isProcess64Bit: boolean = process.arch === 'x64';
+const isOS64Bit: boolean = isProcess64Bit || env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+
+export interface IPowerShellExeDetails {
+	readonly displayName: string;
+	readonly exePath: string;
+}
+
+export interface IPossiblePowerShellExe extends IPowerShellExeDetails {
+	exists(): boolean;
+}
+
+class PossiblePowerShellExe implements IPossiblePowerShellExe {
+	public readonly exePath: string;
+	public readonly displayName: string;
+
+	private knownToExist: boolean | undefined;
+
+	constructor(
+		pathToExe: string,
+		installationName: string,
+		{ knownToExist = false }: { knownToExist?: boolean } = {}) {
+
+		this.exePath = pathToExe;
+		this.displayName = installationName;
+		this.knownToExist = knownToExist || undefined;
+	}
+
+	public exists(): boolean {
+		if (this.knownToExist === undefined) {
+			this.knownToExist = fs.existsSync(this.exePath);
+		}
+		return this.knownToExist;
+	}
+}
+
+function getProgramFilesPath(
+	{ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string | null {
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return env.ProgramFiles!;
+	}
+
+	// We might be a 64-bit process looking for 32-bit program files
+	if (isProcess64Bit) {
+		return env['ProgramFiles(x86)']!;
+	}
+
+	// We might be a 32-bit process looking for 64-bit program files
+	if (isOS64Bit) {
+		return env.ProgramW6432!;
+	}
+
+	// We're a 32-bit process on 32-bit Windows, there is no other Program Files dir
+	return null;
+}
+
+function getSystem32Path({ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string {
+	const windir: string = env.windir!;
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return path.join(windir, 'System32');
+	}
+
+	// We might be a 64-bit process looking for 32-bit system32
+	if (isProcess64Bit) {
+		return path.join(windir, 'SysWOW64');
+	}
+
+	// We might be a 32-bit process looking for 64-bit system32
+	if (isOS64Bit) {
+		return path.join(windir, 'Sysnative');
+	}
+
+	// We're on a 32-bit Windows, so no alternate bitness
+	return path.join(windir, 'System32');
+}
+
+function findPSCoreWindowsInstallation(
+	{ useAlternateBitness = false, findPreview = false }:
+		{ useAlternateBitness?: boolean; findPreview?: boolean } = {}): IPossiblePowerShellExe | null {
+
+	const programFilesPath = getProgramFilesPath({ useAlternateBitness });
+	if (!programFilesPath) {
+		return null;
+	}
+
+	const powerShellInstallBaseDir = path.join(programFilesPath, 'PowerShell');
+
+	// Ensure the base directory exists
+	if (!(fs.existsSync(powerShellInstallBaseDir) && fs.lstatSync(powerShellInstallBaseDir).isDirectory())) {
```

Comment:
`vscode.env.shell` eventually calls down to this method... "promisifying" this API might be quite a lot of work. Should I still do it? I did move over to `pfs` for everything, but there it uses sync `fs` APIs... Let me know what you think is best.

Reply:
I don't think we should try to promisify the extension API. Does this need to totally reevaluate every time something calls `getSystemShell` or would it be ok to run once async, cache the result, and be able to return it synchronously later? If that won't work, maybe we would still want an async version to use internally. The sync fs checking would run during startup and block other stuff cc @bpasero

Reply:
Yes ideally should not use sync fs API anywhere in startup. To be fair though, we have a bit of sync code on startup simply because on startup nothing else executes and we never pushed async fs all the way through there. But it is not recommended (https://github.com/microsoft/vscode/issues/16381, https://github.com/microsoft/vscode/issues/16387)

Reply:
For what it's worth, the default shell is cached so those sync calls would only get called once at startup... 

They would, however, get called every time the user wants to change their shell (how often would that be? Also if they do this follow, then afterwards the startup code pulls from the setting and doesn't run any of these sync calls, I believe). I could see about having an async version for just this scenario and leave the sync ones for discovering the default at startup. What do you think of that plan?

If not, any idea where the async code could go so that this getDefaultShell method can stay sync? I don't think we have the option of making that async because that would be a breaking change for `vscode.env.shell`.

Reply:
Adding @Tyriar as well in case he has any ideas.

Reply:
Oh, I just commented the same thing before reading this 😆

Reply:
See https://github.com/microsoft/vscode/pull/112768#discussion_r551997816